[16.0][IMP] sale_loyalty_order_suggestion: Add hook method

pilarvargas-tecnativa opened the pull request:
Add hook method to help communicate this module with other modules that extend it by returning records from programs whose rules are based on products before filtering by order product and avoid extra modules that will not provide any functionality.

cc @Tecnativa TT48349

@chienandalu @pedrobaeza please review

pilarvargas-tecnativa committed:
[IMP] sale_loyalty_order_suggestion: Add hook method

Add hook method to help communicate this module with other modules that
extend it by returning records from programs whose rules are based on
products before filtering by order product and avoid extra modules that
will not provide any functionality.

TT48349

pedrobaeza commented on sale_loyalty_order_suggestion/models/sale_order.py:
```diff
@@ -73,17 +73,24 @@ def _get_available_programs(self):
+        filtering that corresponds to each functionality."""
         valid_programs = self._get_available_programs()
         # Filters programs that have rules with minimum_qty > 0
         programs_with_minimum_qty = valid_programs.filtered(
```
Why this module includes this minimum quantity criteria?

pilarvargas-tecnativa replied:
The purpose of this module is to suggest promotions on order lines whose product is part of the rules of a promotion.

When the suggestion is made, it shows a gift icon which, when clicked, allows us to configure what is necessary for the promotion to be applicable (products and quantity to be added to the sales order).

By default, promotions in their rules consider the domain [ ] for the allowed products if a specific product is not set in their rules.

The rules of a promotion are not restrictive to each other but as soon as one is fulfilled, the promotion is applicable.

For this reason, and following the example of the DEMO promotion "Buy 3 large cabinets, get one for free" of the type "Buy X get Y", for a promotion to have its rules based on products, it is necessary to establish a quantity as the main criterion.

![image](https://github.com/OCA/sale-promotion/assets/118818446/92b1c7b0-bfce-43eb-bf9f-aac1751fdf7f)

The rest of the types of promotions can also base their rules on products and that is why in this logic they are filtered by product quantity.

It should be clarified that the different types of promotions facilitate a more direct configuration (giving the configuration practically done) of the rules, but in reality we can configure anything as we need.

pilarvargas-tecnativa committed:
[IMP] sale_loyalty_order_suggestion: Add product_id field to the wizard

Since we cannot know which rule is being approved for the promotion to be
applicable, in case of multi-rules it may be necessary to have available
the product that makes the promotion suggestion in order to filter the
first rule that complies with that product.

pedrobaeza commented:
/ocabot merge patch
